### DISCUSSION TITLE
How to access PDO?

### DISCUSSION THREAD

**Discussion by murraycollingwood** [propelorm/Propel2#1867] [Q&A] [answered]
Just upgraded to Propel2 and I'm having trouble accessing the PDO object.

The main issue is that I have a Sabre library (for processing CalDAV requests) that requires the PDO connection to my database as a parameter.
```
$server = new Sabre_CalDAV_Server($pdo, $public = true);
```

Initially I had used this option:
```
/* Database */
$pdo = Propel\Runtime\Propel::getReadConnection(SOBS_DATASOURCE);
$pdo->setAttribute(PDO::ATTR_ERRMODE,PDO::ERRMODE_EXCEPTION);
```

However that was passing a ConnectionWrapper object - not what I was after.
So I googled a whole lot and found this option:
```
/* Database */
$pdo = new \Propel\Runtime\Connection\PdoConnection('mysql:host=localhost;dbname=sobs', 'username', 'password');
$pdo->setAttribute(PDO::ATTR_ERRMODE,PDO::ERRMODE_EXCEPTION);
```

However even with this option I'm still getting an error, like this:
```
[13-4-22 5:52pm] [0]: Uncaught exception "TypeError" with message "Argument 1 passed to Sabre_CalDAV_Server::__construct() must be an instance of PDO, instance of Propel\Runtime\Connection\PdoConnection given
```

How do I extract the actual PDO object from Propel that I can pass to this function?

Cheers
Murray

**Comment by mringler** [comment, marked as answer]
Oh, I don't think this is possible at the moment.
In your case you would want Propel to create another connection (which it is not using itself) and then hand it to you. While I see how this would be convenient, it is currently not implemented from what I see. I also don't see a way how to access the connection information (host, user, password, etc).
So I guess easiest approach is to create a new connection yourself, from connection information you get from outside Propel.

If you feel like using Propel to create PDO connections is a missing feature and want to create a PR, I think it would be interesting.

Best,
Moritz

**Reply by murraycollingwood** [reply]
Thanks Moritz, this was my thought also.  I had to create a separate PDO connection, which meant entering my 'dsn' username and password is now in two places.  That's not ideal.
Is there a way of retrieving the dsn, username and password from the Propel connection object to support this (so that I only have it specified in one place)?

Cheers
Murray

**Reply by mringler** [reply]
Hey Murray,
you could try something like this:
```php
       $sc = Propel::getServiceContainer();
       $connectionName = 'default'; // check in your config.php for connection name
       $connectionManager = $sc->getConnectionManager($connectionName);
       $configuration = $connectionManager->getConfiguration();
       ['dsn' => $dsn, 'user' => $user, 'password' => $password] = $configuration;
```
If you use a type checker, it will complain since `getConnectionManager()` returns a ConnectionManagerInterface, which does not expose `getConfiguration()`, but it is a public method on ConnectionManagerSingle, which you will most likely get.

Hope that helps, best, 
Moritz
